--- a/ipc/namespace.c
+++ b/ipc/namespace.c
@@ -33,10 +33,10 @@ static struct ipc_namespace *create_ipc_ns(struct user_namespace *user_ns,
 	ns->ns.ops = &ipcns_operations;
 
 	atomic_set(&ns->count, 1);
+	ns->user_ns = get_user_ns(user_ns);
 	err = mq_init_ns(ns);
 	if (err) {
+		put_user_ns(ns->user_ns);
 		ns_free_inum(&ns->ns);
 		kfree(ns);
 		return ERR_PTR(err);
@@ -46,7 +46,6 @@ static struct ipc_namespace *create_ipc_ns(struct user_namespace *user_ns,
 	sem_init_ns(ns);
 	msg_init_ns(ns);
 	shm_init_ns(ns);
-
-	ns->user_ns = get_user_ns(user_ns);
 
 	return ns;
 }
